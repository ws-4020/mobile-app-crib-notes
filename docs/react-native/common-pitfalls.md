---
title: ハマりがちな落とし穴
sidebar_label: Overview
---

React Nativeで開発する際に起きた困り事や不具合についてまとめています。
