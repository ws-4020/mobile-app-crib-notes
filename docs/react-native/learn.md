---
title: About
hide_table_of_contents: true
---

このコンテンツはReact Nativeでのモバイルアプリケーション開発を学習するためのコンテンツです。
