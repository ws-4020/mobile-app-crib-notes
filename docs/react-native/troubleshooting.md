---
title: トラブルシュート
---

React Nativeで開発する中で起きた困り事や不具合についてまとめています。
