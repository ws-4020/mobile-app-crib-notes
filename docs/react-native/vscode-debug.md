---
title: Visual Studio Codeの設定
---

ここではVisual Studio Code (以下VS Code) での開発に際して有用なプラグインをご紹介します。

## React Native Tools

VS Codeプラグインである`React Native Tools`を利用することで開発中のモバイルアプリケーションをデバッグできます。

VS Codeで、React NativeのTypeScriptコードをデバッグできるように、[React Native Tools](https://marketplace.visualstudio.com/items?itemName=msjsdiag.vscode-react-native)をセットアップします。


### React Native Toolsのインストール

VS CodeのReact Nativeの拡張ツールである、`React Native Tools`プラグインをインストールします。

Visual Studioマーケットプレイスの[React Native Toolsページ](https://marketplace.visualstudio.com/items?itemName=msjsdiag.vscode-react-native)の`Install`をクリックしてインストール。

または、VS Codeの「拡張機能」で`React Native Tools`検索してインストール。

### 起動構成を追加する

プロジェクトルート配下の`/.vscode/launch.json`に以下を追加してください。ファイルが無い場合は新規に作成してください。

```json title="/.vscode/launch.json"
{
  "configurations": [
    {
      "name": "Attach to packager",
      "cwd": "${workspaceFolder}",
      "type": "reactnative",
      "request": "attach",
      "port": "19000"
    }
  ]
}
```

:::info
portにはMetro Bundlerのポート番号を指定してください。デフォルトはExpoの場合19000、Expoを使用しない場合は8081です。

ポート番号をデフォルトから変更している場合はその番号を指定してください。
:::

### デバッグ

#### 実機の場合

1. 任意の起動方法にてアプリ起動した後に端末をシェイクし、開発者メニューを表示。
2. `Debug Remote JS`を選択した後、ブラウザで`debugger-ui`が立ち上がるのでそのタブを閉じる。
3. VS Codeの「実行とデバッグ」から`Attach to packager`をクリックして実行。
4. 開発者メニューからアプリをリロード。

#### シミュレータの場合

##### iOSシミュレータ

1. 任意の起動方法にてアプリ起動した後に`Ctrl+cmd+z`を押下し、開発者メニューを表示。
2. `Debug Remote JS`を選択した後、ブラウザで`debugger-ui`が立ち上がるのでそのタブを閉じる。
3. VS Codeの「実行とデバッグ」から`Attach to packager`をクリックして実行。
4. 開発者メニューからアプリをリロード。

##### Androidシミュレータ

1. 任意の起動方法にてアプリ起動した後にWindows環境の場合`Ctrl+m`、 Mac環境の場合`cmd+m`を押下し、開発者メニューを表示。
2. `Debug Remote JS`を選択した後、ブラウザで`debugger-ui`が立ち上がるのでそのタブを閉じる。
3. VS Codeの「実行とデバッグ」から`Attach to packager`をクリックして実行。
4. 開発者メニューからアプリをリロード。

## Lint、フォーマッタ

- [Prettier - Code formatter](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode)
- [ESLint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint)

## その他

- [GitLens — Git supercharged](https://marketplace.visualstudio.com/items?itemName=eamodio.gitlens)
- [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker)
- [Path Intellisense](https://marketplace.visualstudio.com/items?itemName=christian-kohler.path-intellisense)
- [Color Highlight](https://marketplace.visualstudio.com/items?itemName=naumovs.color-highlight)
- [Rainbow Brackets](https://marketplace.visualstudio.com/items?itemName=2gua.rainbow-brackets)
- [Markdown All in One](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one)
