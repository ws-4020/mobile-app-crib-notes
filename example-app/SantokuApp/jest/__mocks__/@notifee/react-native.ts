/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import notifee from '@notifee/react-native';
export * from '@notifee/react-native/dist/types/Library';
export * from '@notifee/react-native/dist/types/Notification';
export * from '@notifee/react-native/dist/types/Trigger';
export * from '@notifee/react-native/dist/types/NotificationIOS';
export * from '@notifee/react-native/dist/types/NotificationAndroid';

const mock: typeof notifee = {
  cancelAllNotifications: jest.fn(),
  cancelDisplayedNotifications: jest.fn(),
  cancelTriggerNotifications: jest.fn(),
  cancelNotification: jest.fn(),
  cancelDisplayedNotification: jest.fn(),
  cancelTriggerNotification: jest.fn(),
  createChannel: jest.fn(),
  createChannels: jest.fn(),
  createChannelGroup: jest.fn(),
  createChannelGroups: jest.fn(),
  deleteChannel: jest.fn(),
  deleteChannelGroup: jest.fn(),
  displayNotification: jest.fn(),
  openAlarmPermissionSettings: jest.fn(),
  createTriggerNotification: jest.fn(),
  getTriggerNotificationIds: jest.fn(),
  getDisplayedNotifications: jest.fn(),
  getTriggerNotifications: jest.fn(),
  getChannel: jest.fn(),
  isChannelCreated: jest.fn(),
  isChannelBlocked: jest.fn(),
  getChannels: jest.fn(),
  getChannelGroup: jest.fn(),
  getChannelGroups: jest.fn(),
  getInitialNotification: jest.fn(),
  onBackgroundEvent: jest.fn(),
  onForegroundEvent: jest.fn(),
  openNotificationSettings: jest.fn(),
  registerForegroundService: jest.fn(),
  stopForegroundService: jest.fn(),
  requestPermission: jest.fn(),
  setNotificationCategories: jest.fn(),
  getNotificationCategories: jest.fn(),
  getNotificationSettings: jest.fn(),
  getBadgeCount: jest.fn(),
  setBadgeCount: jest.fn(),
  incrementBadgeCount: jest.fn(),
  decrementBadgeCount: jest.fn(),
  openBatteryOptimizationSettings: jest.fn(),
  isBatteryOptimizationEnabled: jest.fn(),
  getPowerManagerInfo: jest.fn(),
  openPowerManagerSettings: jest.fn(),
  hideNotificationDrawer: jest.fn(),
  SDK_VERSION: '5.4.1',
};

// 複数のファイルでnotifeeをimportしていた場合に、redefineされないようにモックが存在するか確認
if (!__mocks.notifee) {
  Object.defineProperty(__mocks, 'notifee', {value: mock});
}

export default mock;
