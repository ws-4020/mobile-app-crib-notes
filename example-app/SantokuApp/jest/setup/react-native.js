/**
 * Copyright 2023 TIS Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// https://reactnavigation.org/docs/testing#mocking-native-modules
// Silence the warning: Animated: `useNativeDriver` is not supported because the native animated module is missing
jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper');

// !!! NOTICE !!!
// The React Navigation documentation says that `react-native-reanimated` and `react-native-gesture-handler`
// also need to be configured with mocks.
// But since React Navigation v6 does not already depend on those packages, mocking those packages is not necessary
// for React Navigation v6.
//
// If you want to use `react-native-reanimated` and `react-native-gesture-handler`, add mock according to the guide
// provided by each package.
// - `react-native-reanimated`: https://docs.swmansion.com/react-native-reanimated/docs/guide/testing/
// - `react-native-gesture-handler`: https://docs.swmansion.com/react-native-gesture-handler/docs/guides/testing
