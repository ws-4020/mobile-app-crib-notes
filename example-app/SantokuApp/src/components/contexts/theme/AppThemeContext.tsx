import React, {createContext} from 'react';
import {useColorScheme} from 'react-native-appearance';
import {ThemeProvider} from 'react-native-elements';

import {AppTheme, darkModeAppTheme, lightModeAppTheme} from './AppTheme';
import {getReactNativeElementsTheme} from './ReactNativeElementsTheme';

const defaultAppTheme = lightModeAppTheme;
const AppThemeContext = createContext<AppTheme>(defaultAppTheme);

const AppThemeProvider: React.FC = ({children}) => {
  const colorScheme = useColorScheme();
  const appTheme = colorScheme === 'dark' ? darkModeAppTheme : lightModeAppTheme;
  const reactNativeElementsTheme = getReactNativeElementsTheme(appTheme);

  return (
    <AppThemeContext.Provider value={appTheme}>
      <ThemeProvider theme={reactNativeElementsTheme}>{children}</ThemeProvider>
    </AppThemeContext.Provider>
  );
};

export {AppThemeContext, AppThemeProvider};
