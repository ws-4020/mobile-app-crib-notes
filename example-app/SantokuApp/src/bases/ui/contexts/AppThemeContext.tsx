import React, {useMemo} from 'react';
import {ColorSchemeName, useColorScheme} from 'react-native';
import {ThemeProvider} from 'react-native-elements';

import {darkModeAppTheme, lightModeAppTheme} from '../constants/AppTheme';
import {getReactNativeElementsTheme} from '../utils/ReactNativeElementsTheme';
import {AppThemeContextProvider} from './useAppTheme';

const getAppTheme = (colorScheme: ColorSchemeName) => (colorScheme === 'dark' ? darkModeAppTheme : lightModeAppTheme);

const WithAppTheme: React.FC = ({children}) => {
  const colorScheme = useColorScheme();
  const appTheme = getAppTheme(colorScheme);
  const reactNativeElementsTheme = useMemo(() => getReactNativeElementsTheme(appTheme), [appTheme]);

  return (
    <AppThemeContextProvider value={appTheme}>
      <ThemeProvider theme={reactNativeElementsTheme}>{children}</ThemeProvider>
    </AppThemeContextProvider>
  );
};

export {WithAppTheme};
