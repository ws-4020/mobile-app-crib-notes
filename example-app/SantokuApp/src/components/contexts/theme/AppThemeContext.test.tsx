import '@testing-library/jest-native/extend-expect';
import {render, RenderOptions} from '@testing-library/react-native';
import React, {useContext} from 'react';
import {Text} from 'react-native';
import {AppearanceProvider, useColorScheme} from 'react-native-appearance';

import {darkModeAppTheme, lightModeAppTheme} from './AppTheme';
import {AppThemeContext, AppThemeProvider} from './AppThemeContext';

const renderWithAppearanceProvider = (component: React.ReactElement<any>, options?: RenderOptions) => {
  return render(component, {wrapper: AppearanceProvider, ...options});
};

jest.mock('react-native-appearance', () => {
  return {
    useColorScheme: jest.fn().mockName('mockUseColorScheme').mockReturnValue('light'),
  };
});

describe('AppThemeProvider', () => {
  const ChildComponent: React.FC = () => {
    const appTheme = useContext(AppThemeContext);
    return <Text testID="test">{appTheme.name}</Text>;
  };

  it('AppThemeProviderを子要素を含めて正常にrenderできること', () => {
    const appThemeProvider = renderWithAppearanceProvider(
      <AppThemeProvider>
        <ChildComponent />
      </AppThemeProvider>,
    );
    expect(appThemeProvider.queryByTestId('test')).not.toBeNull();
  });

  it('colorSchemeがlightの場合にlight用のAppThemeが取得できること', () => {
    const mockUseColorScheme = useColorScheme as jest.Mock;
    mockUseColorScheme.mockReturnValueOnce('light');
    const appThemeProvider = renderWithAppearanceProvider(
      <AppThemeProvider>
        <ChildComponent />
      </AppThemeProvider>,
    );

    expect(mockUseColorScheme).toBeCalled();
    expect(appThemeProvider.getByText(lightModeAppTheme.name)).not.toBeNull();
    expect(appThemeProvider).toMatchSnapshot();
  });

  it('colorSchemeがdarkの場合にdark用のAppThemeが取得できること', () => {
    const mockUseColorScheme = useColorScheme as jest.Mock;
    mockUseColorScheme.mockReturnValueOnce('dark');
    const appThemeProvider = renderWithAppearanceProvider(
      <AppThemeProvider>
        <ChildComponent />
      </AppThemeProvider>,
    );

    expect(mockUseColorScheme).toBeCalled();
    expect(appThemeProvider.getByText(darkModeAppTheme.name)).not.toBeNull();
    expect(appThemeProvider).toMatchSnapshot();
  });
});
