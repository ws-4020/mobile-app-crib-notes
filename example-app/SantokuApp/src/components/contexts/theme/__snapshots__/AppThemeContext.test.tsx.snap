// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemeProvider colorSchemeがdarkの場合にdark用のAppThemeが取得できること 1`] = `
<Text
  testID="test"
>
  darkTheme
</Text>
`;

exports[`AppThemeProvider colorSchemeがlightの場合にlight用のAppThemeが取得できること 1`] = `
<Text
  testID="test"
>
  lightTheme
</Text>
`;
